File: src/g/config_validation.go
package g

import (
	"errors"
	"fmt"
	"math"
	"net"
	"net/url"
	"strconv"
	"strings"
)

const (
	maxConfigTimeoutSeconds = 60
	maxConfigPort           = 65535
	maxConfigNetworkNodes   = 1024
	maxConfigTargetsPerNode = 1024
	maxConfigMappingTargets = 8192
	maxConfigAuthorizedIPs  = 4096
)

func ValidateConfig(config Config) error {
	if strings.TrimSpace(config.Name) == "" {
		return errors.New("本机节点名称为空!")
	}
	if config.Port < 1 || config.Port > maxConfigPort {
		return fmt.Errorf("非法监听端口!(1-%d)", maxConfigPort)
	}
	if !validIPv4(config.Addr) {
		return errors.New("非法本机节点IP!")
	}
	if config.Base["Timeout"] < 1 || config.Base["Timeout"] > maxConfigTimeoutSeconds {
		return fmt.Errorf("非法超时时间!(1-%d秒)", maxConfigTimeoutSeconds)
	}
	if config.Base["Archive"] < 1 || config.Base["Archive"] > 36500 {
		return errors.New("非法存档天数!(1-36500天)")
	}
	if config.Base["Refresh"] < 1 || config.Base["Refresh"] > 1440 {
		return errors.New("非法刷新频率!(1-1440分钟)")
	}
	optionalLimits := map[string][2]int{
		"PingCount":          {1, 120},
		"PingIntervalMs":     {100, 60000},
		"PingTimeoutMs":      {100, 60000},
		"PingStaggerMs":      {0, 60000},
		"MappingConcurrency": {1, 64},
		"MappingProbeCount":  {1, 20},
	}
	for key, limits := range optionalLimits {
		if value, ok := config.Base[key]; ok && (value < limits[0] || value > limits[1]) {
			return fmt.Errorf("非法配置参数 %s!(%d-%d)", key, limits[0], limits[1])
		}
	}

	if err := validatePositiveFloat(config.Topology["Tline"], 20, "非法拓扑连线粗细!(0-20)"); err != nil {
		return err
	}
	if err := validatePositiveFloat(config.Topology["Tsymbolsize"], 500, "非法拓扑形状大小!(0-500)"); err != nil {
		return err
	}
	if config.Toollimit < 0 || config.Toollimit > 86400 {
		return errors.New("非法检测工具限定频率!(0-86400秒)")
	}
	if len(config.Network) == 0 {
		return errors.New("Ping节点测试网络信息为空!")
	}
	if len(config.Network) > maxConfigNetworkNodes {
		return fmt.Errorf("Ping节点数量超过上限!(%d)", maxConfigNetworkNodes)
	}
	if _, ok := config.Network[config.Addr]; !ok {
		return errors.New("本机节点未包含在网络配置中!")
	}

	for key, member := range config.Network {
		if !validIPv4(key) || !validIPv4(member.Addr) || key != member.Addr {
			return fmt.Errorf("Ping节点测试网络信息错误!(节点键与地址不一致 %s)", key)
		}
		if strings.TrimSpace(member.Name) == "" {
			return fmt.Errorf("Ping节点测试网络信息错误!(%s 节点名称为空)", key)
		}
		if len(member.Ping) > maxConfigTargetsPerNode || len(member.Topology) > maxConfigTargetsPerNode {
			return fmt.Errorf("Ping节点测试网络信息错误!(%s 目标数量超过上限 %d)", key, maxConfigTargetsPerNode)
		}
		seenPingTargets := make(map[string]struct{}, len(member.Ping))
		for _, target := range member.Ping {
			if !validIPv4(target) {
				return fmt.Errorf("Ping节点测试网络信息错误!(非法目标IP %s)", target)
			}
			if _, ok := config.Network[target]; !ok {
				return fmt.Errorf("Ping节点测试网络信息错误!(目标节点不存在 %s)", target)
			}
			if _, exists := seenPingTargets[target]; exists {
				return fmt.Errorf("Ping节点测试网络信息错误!(重复目标节点 %s)", target)
			}
			seenPingTargets[target] = struct{}{}
		}
		seenTopologyTargets := make(map[string]struct{}, len(member.Topology))
		for _, topology := range member.Topology {
			if err := validateTopologyRule(key, topology, config.Network); err != nil {
				return err
			}
			target := topology["Addr"]
			if _, exists := seenTopologyTargets[target]; exists {
				return fmt.Errorf("Ping节点测试网络信息错误!(%s 重复拓扑目标 %s)", key, target)
			}
			seenTopologyTargets[target] = struct{}{}
		}
	}

	mappingTargets := 0
	allowedMappingCarriers := map[string]struct{}{"ctcc": {}, "cucc": {}, "cmcc": {}}
	for province, providers := range config.Chinamap {
		for carrier, ips := range providers {
			if _, ok := allowedMappingCarriers[carrier]; !ok {
				return fmt.Errorf("非法 Mapping 运营商!(%s: %s)", province, carrier)
			}
			mappingTargets += len(ips)
			if mappingTargets > maxConfigMappingTargets {
				return fmt.Errorf("Mapping IP数量超过上限!(%d)", maxConfigMappingTargets)
			}
			for _, ip := range ips {
				if ip != "" && !validIPv4(ip) {
					return errors.New("Mapping Ip illegal!")
				}
			}
		}
	}
	if err := validateAuthIPList(config.Authiplist); err != nil {
		return err
	}
	return validateMode(config.Mode)
}

func validIPv4(value string) bool {
	if value != strings.TrimSpace(value) {
		return false
	}
	ip := net.ParseIP(value)
	return ip != nil && ip.To4() != nil
}

func validatePositiveFloat(raw string, max float64, message string) error {
	value, err := strconv.ParseFloat(raw, 64)
	if err != nil || math.IsNaN(value) || math.IsInf(value, 0) || value <= 0 || value > max {
		return errors.New(message)
	}
	return nil
}

func validateTopologyRule(source string, rule map[string]string, network map[string]NetworkMember) error {
	target := rule["Addr"]
	if !validIPv4(target) {
		return fmt.Errorf("Ping节点测试网络信息错误!(%s 非法拓扑目标 %s)", source, target)
	}
	if _, ok := network[target]; !ok {
		return fmt.Errorf("Ping节点测试网络信息错误!(%s 拓扑目标不存在 %s)", source, target)
	}
	if strings.TrimSpace(rule["Name"]) == "" {
		return fmt.Errorf("Ping节点测试网络信息错误!(%s->%s 拓扑名称为空)", source, target)
	}
	limits := map[string][2]int{
		"Thdchecksec": {1, 86400},
		"Thdloss":     {0, 100},
		"Thdavgdelay": {1, 60000},
		"Thdoccnum":   {1, 10000},
	}
	parsedValues := make(map[string]int, len(limits))
	for key, bounds := range limits {
		value, err := strconv.Atoi(rule[key])
		if err != nil || value < bounds[0] || value > bounds[1] {
			return fmt.Errorf("Ping节点测试网络信息错误!(%s->%s 非法报警规则 %s)", source, target, key)
		}
		if key == "Thdchecksec" && value%60 != 0 {
			return fmt.Errorf("Ping节点测试网络信息错误!(%s->%s 报警检查窗口必须为60秒的倍数)", source, target)
		}
		parsedValues[key] = value
	}
	maxOccurrences := parsedValues["Thdchecksec"] / 60
	if parsedValues["Thdoccnum"] > maxOccurrences {
		return fmt.Errorf("Ping节点测试网络信息错误!(%s->%s 报警发生次数不能超过检查窗口内的样本数 %d)", source, target, maxOccurrences)
	}
	return nil
}

func validateAuthIPList(raw string) error {
	values := strings.Split(strings.ReplaceAll(raw, " ", ""), ",")
	if len(values) > maxConfigAuthorizedIPs {
		return fmt.Errorf("授权IP数量超过上限!(%d)", maxConfigAuthorizedIPs)
	}
	for _, value := range values {
		if value != "" && net.ParseIP(value) == nil {
			return fmt.Errorf("非法授权IP地址: %s", value)
		}
	}
	return nil
}

func validateMode(mode map[string]string) error {
	modeType := mode["Type"]
	if modeType == "" || modeType == "local" {
		return nil
	}
	if modeType != "cloud" {
		return errors.New("非法运行模式!")
	}
	endpoint, err := url.Parse(mode["Endpoint"])
	if err != nil || !validCloudEndpoint(endpoint) {
		return errors.New("非法云配置地址!")
	}
	return nil
}

func validCloudEndpoint(endpoint *url.URL) bool {
	if endpoint == nil ||
		!endpoint.IsAbs() ||
		(endpoint.Scheme != "http" && endpoint.Scheme != "https") ||
		endpoint.Hostname() == "" ||
		endpoint.User != nil ||
		endpoint.Fragment != "" {
		return false
	}
	port := endpoint.Port()
	if port == "" {
		return true
	}
	value, err := strconv.Atoi(port)
	return err == nil && value >= 1 && value <= maxConfigPort
}

File: web/src/utils/configValidation.ts
import type { Config, TopologyConfig } from '@/types'

export const CONFIG_LIMITS = {
  timeout: { min: 1, max: 60 },
  archive: { min: 1, max: 36500 },
  refresh: { min: 1, max: 1440 },
  topologyLine: { min: 0, max: 20 },
  topologySymbol: { min: 0, max: 500 },
  toolLimit: { min: 0, max: 86400 },
  networkNodes: 1024,
  targetsPerNode: 1024,
  mappingTargets: 8192,
  authorizedIps: 4096
} as const

const OPTIONAL_BASE_LIMITS = {
  PingCount: { min: 1, max: 120 },
  PingIntervalMs: { min: 100, max: 60000 },
  PingTimeoutMs: { min: 100, max: 60000 },
  PingStaggerMs: { min: 0, max: 60000 },
  MappingConcurrency: { min: 1, max: 64 },
  MappingProbeCount: { min: 1, max: 20 }
} as const

export interface ConfigValidationIssue {
  key: string
  params?: Record<string, number | string>
}

const issue = (key: string, params?: Record<string, number | string>): ConfigValidationIssue => ({
  key,
  params
})

const isIntegerInRange = (value: number, min: number, max: number) =>
  Number.isInteger(value) && value >= min && value <= max

const isPositiveFloatInRange = (value: string, max: number) => {
  if (value.trim() === '' || value !== value.trim()) {
    return false
  }
  const parsed = Number(value)
  return Number.isFinite(parsed) && parsed > 0 && parsed <= max
}

const parseInteger = (value: string): number | null => {
  if (!/^[+-]?\d+$/.test(value)) {
    return null
  }
  const parsed = Number(value)
  return Number.isSafeInteger(parsed) ? parsed : null
}

export const isValidIPv4 = (value: string): boolean => {
  if (value !== value.trim()) {
    return false
  }
  const parts = value.split('.')
  return (
    parts.length === 4 &&
    parts.every(
      (part) => /^\d{1,3}$/.test(part) && Number(part) <= 255 && String(Number(part)) === part
    )
  )
}

const isValidIPv6 = (value: string): boolean => {
  if (value !== value.trim() || !value.includes(':')) {
    return false
  }

  let normalized = value
  if (normalized.includes('.')) {
    const lastColon = normalized.lastIndexOf(':')
    const ipv4 = normalized.slice(lastColon + 1)
    if (lastColon < 0 || !isValidIPv4(ipv4)) {
      return false
    }
    normalized = `${normalized.slice(0, lastColon)}:0:0`
  }

  if (!/^[0-9a-f:]+$/i.test(normalized) || (normalized.match(/::/g) ?? []).length > 1) {
    return false
  }

  const hasCompression = normalized.includes('::')
  const [left = '', right = ''] = normalized.split('::')
  const leftParts = left ? left.split(':') : []
  const rightParts = right ? right.split(':') : []
  const parts = [...leftParts, ...rightParts]
  if (parts.some((part) => !/^[0-9a-f]{1,4}$/i.test(part))) {
    return false
  }
  return hasCompression ? parts.length < 8 : parts.length === 8
}

const isValidIPAddress = (value: string) => isValidIPv4(value) || isValidIPv6(value)

const validateTopologyRule = (
  source: string,
  rule: TopologyConfig,
  config: Config
): ConfigValidationIssue | null => {
  if (!isValidIPv4(rule.Addr) || !config.Network[rule.Addr] || !rule.Name.trim()) {
    return issue('config.validationTopologyTarget', { source, target: rule.Addr || '-' })
  }

  const checkSeconds = parseInteger(rule.Thdchecksec)
  const loss = parseInteger(rule.Thdloss)
  const avgDelay = parseInteger(rule.Thdavgdelay)
  const occurrences = parseInteger(rule.Thdoccnum)
  if (
    checkSeconds === null ||
    !isIntegerInRange(checkSeconds, 1, 86400) ||
    checkSeconds % 60 !== 0 ||
    loss === null ||
    !isIntegerInRange(loss, 0, 100) ||
    avgDelay === null ||
    !isIntegerInRange(avgDelay, 1, 60000) ||
    occurrences === null ||
    !isIntegerInRange(occurrences, 1, 10000) ||
    occurrences > checkSeconds / 60
  ) {
    return issue('config.validationTopologyRule', { source, target: rule.Addr })
  }
  return null
}

export const validateConfigForEdit = (config: Config): ConfigValidationIssue | null => {
  if (!config.Name.trim()) {
    return issue('config.validationNodeName')
  }
  if (!isIntegerInRange(config.Port, 1, 65535)) {
    return issue('config.validationPort')
  }
  if (!isValidIPv4(config.Addr)) {
    return issue('config.validationLocalAddress')
  }

  const baseRules = [
    ['config.validationTimeout', config.Base.Timeout, CONFIG_LIMITS.timeout],
    ['config.validationArchive', config.Base.Archive, CONFIG_LIMITS.archive],
    ['config.validationRefresh', config.Base.Refresh, CONFIG_LIMITS.refresh]
  ] as const
  for (const [key, value, limits] of baseRules) {
    if (!isIntegerInRange(value, limits.min, limits.max)) {
      return issue(key, limits)
    }
  }

  for (const [key, limits] of Object.entries(OPTIONAL_BASE_LIMITS)) {
    const value = config.Base[key as keyof typeof OPTIONAL_BASE_LIMITS]
    if (value !== undefined && !isIntegerInRange(value, limits.min, limits.max)) {
      return issue('config.validationBaseParameter', { name: key, ...limits })
    }
  }

  if (!isPositiveFloatInRange(config.Topology.Tline, CONFIG_LIMITS.topologyLine.max)) {
    return issue('config.validationLineWidth')
  }
  if (!isPositiveFloatInRange(config.Topology.Tsymbolsize, CONFIG_LIMITS.topologySymbol.max)) {
    return issue('config.validationSymbolSize')
  }
  if (
    !isIntegerInRange(config.Toollimit, CONFIG_LIMITS.toolLimit.min, CONFIG_LIMITS.toolLimit.max)
  ) {
    return issue('config.validationToolLimit')
  }

  const networkEntries = Object.entries(config.Network)
  if (networkEntries.length === 0) {
    return issue('config.validationNetworkEmpty')
  }
  if (networkEntries.length > CONFIG_LIMITS.networkNodes) {
    return issue('config.validationNetworkLimit', { max: CONFIG_LIMITS.networkNodes })
  }
  if (!config.Network[config.Addr]) {
    return issue('config.validationLocalNodeMissing')
  }

  for (const [address, member] of networkEntries) {
    if (!isValidIPv4(address) || !isValidIPv4(member.Addr) || address !== member.Addr) {
      return issue('config.validationNodeAddress', { address })
    }
    if (!member.Name.trim()) {
      return issue('config.validationNamedNode', { address })
    }
    if (
      member.Ping.length > CONFIG_LIMITS.targetsPerNode ||
      member.Topology.length > CONFIG_LIMITS.targetsPerNode
    ) {
      return issue('config.validationTargetLimit', {
        address,
        max: CONFIG_LIMITS.targetsPerNode
      })
    }

    const seenTargets = new Set<string>()
    for (const target of member.Ping) {
      if (!isValidIPv4(target) || !config.Network[target]) {
        return issue('config.validationPingTarget', { source: address, target })
      }
      if (seenTargets.has(target)) {
        return issue('config.validationDuplicateTarget', { source: address, target })
      }
      seenTargets.add(target)
    }

    for (const topologyRule of member.Topology) {
      const topologyIssue = validateTopologyRule(address, topologyRule, config)
      if (topologyIssue) {
        return topologyIssue
      }
    }
  }

  let mappingTargetCount = 0
  for (const providers of Object.values(config.Chinamap)) {
    for (const addresses of [providers.ctcc, providers.cucc, providers.cmcc]) {
      mappingTargetCount += addresses.length
      if (mappingTargetCount > CONFIG_LIMITS.mappingTargets) {
        return issue('config.validationMappingLimit', { max: CONFIG_LIMITS.mappingTargets })
      }
      const invalidAddress = addresses.find((address) => address !== '' && !isValidIPv4(address))
      if (invalidAddress) {
        return issue('config.validationMappingAddress', { address: invalidAddress })
      }
    }
  }

  const authorizedAddresses = config.Authiplist.replace(/ /g, '').split(',')
  if (authorizedAddresses.length > CONFIG_LIMITS.authorizedIps) {
    return issue('config.validationAuthLimit', { max: CONFIG_LIMITS.authorizedIps })
  }
  const invalidAuthorizedAddress = authorizedAddresses.find(
    (address) => address !== '' && !isValidIPAddress(address)
  )
  if (invalidAuthorizedAddress) {
    return issue('config.validationAuthAddress', { address: invalidAuthorizedAddress })
  }

  const modeType = config.Mode.Type ?? ''
  if (modeType !== '' && modeType !== 'local' && modeType !== 'cloud') {
    return issue('config.validationMode')
  }
  if (modeType === 'cloud') {
    try {
      if (!/^https?:\/\//.test(config.Mode.Endpoint ?? '')) {
        return issue('config.validationCloudEndpoint')
      }
      const endpoint = new URL(config.Mode.Endpoint ?? '')
      if (!['http:', 'https:'].includes(endpoint.protocol) || !endpoint.hostname) {
        return issue('config.validationCloudEndpoint')
      }
    } catch {
      return issue('config.validationCloudEndpoint')
    }
  }

  return null
}
